packages/components/src/internal/components/heatmap/HeatMapDisplay.tsx

```diff
@@ -13,67 +13,68 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import React from 'react';
+import React, { FC, memo, ReactNode, useMemo } from 'react';
 import { List, Map, Set } from 'immutable';
 import { OverlayTrigger, Tooltip } from 'react-bootstrap';
 
-import { HeatMapProps } from './HeatMap';
+import { HeatMapCell, HeatMapDisplayCell, HeatMapProps } from './HeatMap';
 import { naturalSort } from '../../..';
 
+const applySort = (sortFn, labels: Set<string>): Set<string> => {
+    let sorted;
+
+    if (typeof sortFn === 'function') {
+        sorted = labels.sort(sortFn);
+    }
+    else {
+        sorted = labels.sort(naturalSort);
+    }
+
+    return sorted;
+}
+
 interface HeatMapDisplayProps extends HeatMapProps {
-    data: List<Map<string, any>>;
+    data: HeatMapCell[];
     displayMeasure?: boolean;
     onCellClick?: Function;
     onHeaderClick?: Function;
     xSort?: Function;
     ySort?: Function;
-    yAxisColumns: Map<string, any>;
-    yTotalCellRenderer?: (cell: Map<string, any>) => void;
+    yAxisColumns: { [key: string]: HeatMapDisplayCell };
+    yTotalCellRenderer?: (cell: HeatMapDisplayCell) => ReactNode
 }
 
-export class HeatMapDisplay extends React.Component<HeatMapDisplayProps, any> {
-    static applySort(sortFn, labels: Set<any>): Set<any> {
-        let sorted;
-
-        if (typeof sortFn === 'function') {
-            sorted = labels.sort(sortFn);
-        } else {
-            sorted = labels.sort(naturalSort);
-        }
-
-        return sorted;
-    }
-
-    processData() {
-        const { data, xAxis, yAxis, measure, xSort, ySort, onCellClick } = this.props;
+export const HeatMapDisplay: FC<HeatMapDisplayProps> = memo((props) => {
+    const { data, xAxis, yAxis, measure, xSort, ySort, onCellClick, emptyDisplay, onHeaderClick, yTotalCellRenderer,
+        yTotalLabel, yInRangeTotal, yAxisColumns, nounSingular, nounPlural, displayMeasure } = props;
 
+    const processedData = useMemo(() => {
         const rows = [];
-        let xLabels = Set();
-        let yLabels = Set();
-        let measureMap = Map<string, any>();
+        let xLabels = Set<string>();
+        let yLabels = Set<string>();
+        let measureMap = Map<string, HeatMapCell>();
 
         if (!data) {
-            throw new Error('HeatMapDisplay requires a List for data');
+            return;
```

Comment:
No need to have `processedData` return `undefined`. Default to empty values instead so there isn't a need to be defensive when using:

```ts
if (!data) {
    return { headers: [], rows };
}
```

Reply:
yup